### DISCUSSION TITLE
Synchronous SQLAlchemy?

### DISCUSSION THREAD

**Discussion by russdot** [fastapi-users/fastapi-users#1144] [Q&A] [answered]
Is there any way to use non-async SQLAlchemy as the database driver? For example, as the `get_db` dependency injection in the [official FastAPI docs](https://fastapi.tiangolo.com/tutorial/sql-databases/#create-the-database-tables)?  (which uses `sqlalchemy.orm.Session` instead of `sqlalchemy.ext.asyncio.AsyncSession`)

I am attempting to integrate FastAPI Users with [`fastapi-crudrouter`](https://github.com/awtkns/fastapi-crudrouter), and the situation appears to be:
- FastAPI Users uses async SQLAlchemy
- `fastapi-crudrouter`  uses non-async SQLAlchemy (does not yet [support async SQLAlchemy](https://github.com/awtkns/fastapi-crudrouter/issues/122))

Perhaps I'm missing something - but I only see async code in the [SQLAlchemy full example docs](https://fastapi-users.github.io/fastapi-users/10.2/configuration/full-example/#sqlalchemy). Thanks!

**Comment by frankie567** [comment, marked as answer]
Hi @russdot 👋 

Indeed, we only support async SQLAlchemy here. If you want to work with sync SQLAlchemy, you'll need to implement your own database adapter for FastAPI Users. It shall be rather easy, you can copy/paste our official one and adapt it with the standard `sqlalchemy.orm.Session`: https://github.com/fastapi-users/fastapi-users-db-sqlalchemy/blob/686172c77932534e0c78d4b3b7937c997675da9b/fastapi_users_db_sqlalchemy/__init__.py#L83-L179

Basically, you'll likely need to:
* Inject a session of type `sqlalchemy.orm.Session`
* Remove the `await` keyword when working with this session (e.g. `await self.session.commit()` becomes `self.session.commit()`
* The queries shouldn't have to change

**Reply by russdot** [reply]
Thank you for the fast reply and guidance!  I will attempt this soon.

**Reply by jdittrich** [reply]
I tried this but whenever `_get_user ` of the adapted `SQLAlchemyUserDatabase` class is called, FastAPI fails with a "object NoneType can't be used in 'await' expression"-Exception.

**Reply by furkan-guvenc** [reply]
Thanks @frankie567, I created a sync version. It passed all the tests:
[fastapi_users_db_sync_sqlalchemy](https://pypi.org/project/fastapi_users_db_sync_sqlalchemy/5.0.0/)

**Reply by fotinakis** [reply]
I've created a PR for supporting this: https://github.com/fastapi-users/fastapi-users-db-sqlalchemy/pull/13
